primitives.lisp

```diff
@@ -561,18 +562,131 @@ Use the window's resource class.
 Use the window's resource name.
 @end table")
 
-(defstruct frame
-  (number nil :type integer)
-  x
-  y
-  width
-  height
-  window)
-
-(defstruct (head (:include frame))
-  (name "" :type string))
-
-(defclass screen ()
+(defclass swm-class ()
+  ((new-objects
+    :initform nil
+    :accessor swm-class-new-objects
+    :allocation :class
+    :documentation
+"Track all newly created objects in order to mix in the appropriate minor modes
+when they are touched")))
+
+(defmethod initialize-instance :after ((obj swm-class) &key &allow-other-keys)
+  ;; Register all newly created objects so that they can have the relevant minor
+  ;; modes autoenabled.
+  (pushnew obj (swm-class-new-objects obj) :test #'eq))
```

Comment:
Does it make sense to track them in a list? Can't it be an issue when too many objects are tracked? Have you considered using a hash-table instead? You can use a hash-table as a set by using the element you care about as the key and something like t as the value.

Reply:
I think a list is appropriate here because whenever we access the list we walk the entire length to handle all objects before nullifying it. In practice the list seldom grows to more than a handful of elements. When objects are created via `make-swm-class-instance` they have the relevant global minor modes enabled in them, and because they are always towards/at the front of the list grabbing them is performant. If we were to change to hash tables we would be using maphash or similar.